framework/src/utils/SolutionInvalidity.C

```diff
@@ -0,0 +1,147 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SolutionInvalidity.h"
+
+// MOOSE Includes
+#include "MooseError.h"
+#include "MooseApp.h"
+#include "VariadicTable.h"
+
+// System Includes
+#include <chrono>
+#include <memory>
+
+SolutionInvalidity::SolutionInvalidity(MooseApp & app)
+  : ConsoleStreamInterface(app),
+    _solution_invalidity_registry(moose::internal::getSolutionInvalidityRegistry())
+{
+}
+
+SolutionInvalidity::~SolutionInvalidity() {}
+/// Count solution invalid occurrences for each solution id
+void
+SolutionInvalidity::setSolutionInvalid(SolutionID _solution_id)
+{
+  std::lock_guard<std::mutex> lock_id(_invalid_mutex);
+
+  if (_solution_invalid_counts.size() <= _solution_id)
+  {
+    _solution_invalid_counts.resize(_solution_id + 1);
+    ++_solution_invalid_counts[_solution_id];
+  }
+  else
+  {
+    ++_solution_invalid_counts[_solution_id];
+  }
+  _solution_invalid_total_counts.resize(_solution_invalid_counts.size());
+  _solution_invalid_timeiter_counts.resize(_solution_invalid_counts.size());
+}
+
+/// Loop over all the tracked objects and determine whether solution invalid is detected
+bool
+SolutionInvalidity::solutionInvalid() const
+{
+  unsigned int sum_of_elems = 0;
+
+  std::for_each(_solution_invalid_counts.begin(),
+                _solution_invalid_counts.end(),
+                [&](int n) { sum_of_elems += n; });
+  if (sum_of_elems < 1)
+  {
+    return false;
+  }
+  else
+  {
+    return true;
+  }
+}
+
+/// Reset the number of solution invalid occurrences back to zero
+void
+SolutionInvalidity::resetSolutionInvalid()
+{
+  std::fill(_solution_invalid_counts.begin(), _solution_invalid_counts.end(), 0);
+}
+
+/// Reset the number of solution invalid occurrences back to zero
+void
+SolutionInvalidity::resetSolutionInvalidTimeIter()
+{
+  std::fill(_solution_invalid_timeiter_counts.begin(), _solution_invalid_timeiter_counts.end(), 0);
+}
+
+/// Pass the number of solution invalid occurrences from current iteration to comulative counters
+void
+SolutionInvalidity::solutionInvalidAccumulation()
+{
+  for (unsigned int i = 0; i < _solution_invalid_counts.size(); i++)
+  {
+    _solution_invalid_total_counts[i] += _solution_invalid_counts[i];
+    _solution_invalid_timeiter_counts[i] += _solution_invalid_counts[i];
+  }
+}
```

Comment:
All if this will be simplified when we combine it into a single vector.

Reply:
I finished the modification following the example you showed me